### PR TITLE
Implement scheduled evolution orchestrator cycles

### PR DESCRIPTION
**Pull request by ales27pm**
## Summary
- add a background-scheduled evolution orchestrator that checks host idleness, limits VRAM pressure, and rolls out adapters via Ray when enabled
- expose curated self-training dataset collection helpers for orchestration
- extend Ray Serve utilities and add orchestrator-focused unit tests with scheduler and psutil mocks

## Testing
- pytest tests/test_evolution_engine.py -k orchestrator -q

------
https://chatgpt.com/codex/tasks/task_e_68e15a0355548333b33ca6199b76db21

___

### **PR Type**
Enhancement

___

### **Description**
- Add background scheduler for automated evolution training cycles

- Implement resource monitoring with CPU, memory, and VRAM safeguards

- Integrate Ray Serve deployment updates for adapter rollouts

- Add curated dataset collection utilities for self-training

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Background Scheduler"] --> B["Resource Check"]
  B --> C["Dataset Collection"]
  C --> D["MNTP Training"]
  D --> E["Manifest Update"]
  E --> F["Ray Deployment"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>orchestrator.py</strong><dd><code>Background scheduler with resource monitoring</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

modules/evolution_engine/orchestrator.py

<ul><li>Replace simple training pipeline with background scheduler <br>orchestrator<br> <li> Add system resource monitoring (CPU, memory, VRAM) before training<br> <li> Implement automatic Ray Serve deployment updates after training<br> <li> Add model slot management and comprehensive error handling</ul>

</details>

  </td>
  <td><a href="https://github.com/ales27pm/monGARS/pull/176/files#diff-02919c2f4bdf46f70eeb75b2b4e8a098d3b48a5bf214097fc78ba6961915b683">+263/-160</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>self_training.py</strong><dd><code>Curated dataset collection utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

modules/evolution_engine/self_training.py

<ul><li>Add utilities for collecting curated self-training datasets<br> <li> Implement JSONL record parsing with text extraction<br> <li> Support HuggingFace Dataset conversion when available<br> <li> Include dataset catalog integration for latest dataset retrieval</ul>

</details>

  </td>
  <td><a href="https://github.com/ales27pm/monGARS/pull/176/files#diff-2babcb43d4c27b3ad16890b459f70d05f9e38261d16f555b911daa24cbf33340">+117/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>ray_service.py</strong><dd><code>Ray deployment update functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

modules/ray_service.py

<ul><li>Add <code>update_ray_deployment</code> function for runtime adapter updates<br> <li> Implement deployment configuration updates with user config payload<br> <li> Add comprehensive error handling for Ray Serve operations</ul>

</details>

  </td>
  <td><a href="https://github.com/ales27pm/monGARS/pull/176/files#diff-16cfff34ff44e2d5adf93dc2db3cfa9dab9e5c0af44e76e31299994b349e6b16">+34/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_evolution_engine.py</strong><dd><code>Orchestrator unit tests with mocking</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_evolution_engine.py

<ul><li>Add comprehensive orchestrator tests with scheduler mocking<br> <li> Test resource monitoring behavior and training cycle execution<br> <li> Mock system utilities (psutil, torch) for isolated testing<br> <li> Verify Ray deployment rollout integration</ul>

</details>

  </td>
  <td><a href="https://github.com/ales27pm/monGARS/pull/176/files#diff-02ba0230ebb07b6b150cd8c7ffb6351583b618f95d3a1093b9c9343280de16eb">+189/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **New Features**
  * Workflow-driven orchestration for training cycles with configurable scheduling, manual trigger, and graceful shutdown.
  * Resource-aware guards (CPU, memory, VRAM, optional CUDA) and slot coordination to prevent overload.
  * Automated curated data collection/validation, per-run directories, training summaries, artifact/manifest persistence, and optional guarded rollout to live serving.
  * Utilities to collect curated datasets and securely update live deployments with payload validation.

* **Tests**
  * Expanded coverage for scheduling, idle-guard behavior, VRAM scenarios, full training cycles, manifest rollouts, and deployment update validation.

### CODE REVIEW

**Comment by ales27pm** [comment]
@codex fix comments
